### ISSUE TITLE
bdev_get_histogram with buckets

### ISSUE THREAD

**Issue by SWaoter** [spdk/spdk#3543]
# Sighting report

Existing implementation of histograms for bdev is precise, but at the same time is too wide. With default settings histogram covers time ranges up to months. At the same time transferring such histogram requires ~80 kb base64 string.

Let's say that we have few hundreds/thousands bdevs and we want to get their I/O statistics each 30 seconds. It will require:
1) Сall `bdev_get_iostat` (all bdev will be reported in single call). However, `iostat` does not provide needed information about latency. Based on `iostat` we can get only average and max/min(if we call `reset` on each iteration).
2) If we want to get detailed latency information. For instance - calculate percentiles, we have to get histogram.
Right now it requires calling `bdev_get_histogram` for each bdev explicitly and parsing resulting base64.

Described approach will result in very detailed information, but will require addition work from client.
Such detailed information is great for performance testing reports, but for constant monitoring it is too much.

## Possible Solution

Allow client to specify buckets while getting histogram. The result of such call will be not a full histogram, but array of `int`s that represent distribution of histogram into desired buckets. That way it is possible to reduce amount of transferred data per call to `int`*`bucket_count` as well as obtaining detailed information only about needed timeslices. For example request can be like this:

`bdev_get_histogram --name bdev_1 --with-buckets '1us 10us 50us 100us 500us 1s'` . It will result in buckets:
0-1us -> all such requests very fast and client do not need additional separation
1-10us
..
1s - inf -> too slow requests that should not happen at all, so separation between them is also not needed

In addition, because whole histogram report would not be that big, it is possible to include it into `bdev_get_iostat`(not by default, but with option), so whole I/O statistics could be obtained in single call.

Does such observability feature make sense? 

## Context (Environment including OS version, SPDK version, etc.)

Latest SPDK

**Comment by jimharris** [comment]
Hi @SWaoter,

I understand the issues that you are describing. This feature was not designed to support enabling histograms for hundreds or thousands of bdevs all at once, and you are accurately pointing that out here. A couple of observations:

* There is a big memory overhead with the histograms. Each histogram is roughly 58KB `(SHIFT_DEFAULT = 7, (2^7) * (64-7) * sizeof(uint64_t))` in size, and we need to allocate one per bdev per active I/O channel (i.e. possibly one per CPU core). Worst case this is easily GBs of memory for 1000 bdevs. So the bdev_enable_histogram RPC may need ability to specify ability to have a smaller shift value to decrease the memory overhead.
* Everything around the histograms is designed around raw values, usually raw TSC. We avoid expensive div instructions in the I/O path. All post-processing (i.e. translating TSC to microseconds) is done out-of-band. This makes it difficult to allow the user to specify nice round buckets like 500us.

Probably the best way to support something like this would be to allocate a shared memory region that contains the histogram_data structures. Some RPC is used to trigger aggregation of the per-channel histogram_data structures into the one in shared memory. Then separate scripts can be used to generate human-readable histograms from that shared memory region. This avoids marshaling all of this data across JSON-RPC and keeps all of the post-processing logic outside of the main SPDK app.

**Comment by SWaoter** [comment]
Hi @jimharris,

Thank you for your feedback!
Adding ability to specify bucket shift during `bdev_enable_histogram` is definitely useful.
What do you think about adding another size parameter to histogram for specifying number of ranges?
It will enable reducing memory usage by shrinking "long" ranges instead of sacrificing accuracy on all ranges.
It will require adding additional field in `spdk_histogram_data` and modifying `__spdk_histogram_data_get_bucket_range`.
It is extra if for I/O path, but I think that it is not bad tradeoff.

**Comment by jimharris** [comment]
> Hi @jimharris,
> 
> Thank you for your feedback! Adding ability to specify bucket shift during `bdev_enable_histogram` is definitely useful. What do you think about adding another size parameter to histogram for specifying number of ranges? It will enable reducing memory usage by shrinking "long" ranges instead of sacrificing accuracy on all ranges. It will require adding additional field in `spdk_histogram_data` and modifying `__spdk_histogram_data_get_bucket_range`. It is extra if for I/O path, but I think that it is not bad tradeoff.

I am OK with that. Instead of parameter for num ranges though, I would suggest specifying a min and max range index.

Take the example from histogram_data.h:

```
 29 /*
 30  * SPDK histograms are implemented using ranges of bucket arrays.  The most common usage
 31  * model is using TSC datapoints to capture an I/O latency histogram.  For this usage model,
 32  * the histogram tracks only TSC deltas - any translation to microseconds is done by the
 33  * histogram user calling spdk_histogram_data_iterate() to iterate over the buckets to perform
 34  * the translations.
 35  *
 36  * Each range has a number of buckets determined by SPDK_HISTOGRAM_NUM_BUCKETS_PER_RANGE
 37  * which is 128.  The buckets in ranges 0 and 1 each map to one specific datapoint value.
 38  * The buckets in subsequent ranges each map to twice as many datapoint values as buckets
 39  * in the range before it:
 40  *
 41  * Range 0:  1 value each  - 128 buckets cover 0 to 127 (2^7-1)
 42  * Range 1:  1 value each  - 128 buckets cover 128 to 255 (2^8-1)
 43  * Range 2:  2 values each - 128 buckets cover 256 to 511 (2^9-1)
 44  * Range 3:  4 values each - 128 buckets cover 512 to 1023 (2^10-1)
 45  * Range 4:  8 values each - 128 buckets cover 1024 to 2047 (2^11-1)
 46  * Range 5: 16 values each - 128 buckets cover 2048 to 4095 (2^12-1)
 47  * ...
 48  * Range 55: 2^54 values each - 128 buckets cover 2^61 to 2^62-1
 49  * Range 56: 2^55 values each - 128 buckets cover 2^62 to 2^63-1
 50  * Range 57: 2^56 values each - 128 buckets cover 2^63 to 2^64-1
 51  *
 52  * On a 2.3GHz processor, this strategy results in 50ns buckets in the 7-14us range (sweet
 53  * spot for Intel Optane SSD latency testing).
 54  *
 55  * Buckets can be made more granular by increasing SPDK_HISTOGRAM_BUCKET_SHIFT.  This
 56  * comes at the cost of additional storage per namespace context to store the bucket data.
 57  */
```

As you mention, it may often be acceptable to sacrifice those higher range indices. So maybe going up to range 30 is sufficient. That would provide an accurate histogram up to 2^37 TSC, which is roughly 64 seconds (if my math is correct) on a 2GHz processor. Any tallies above that number go into the last bucket in the last range.

But maybe we also don't care about accuracy at the very low end of that range either. On a 2GHz processor, 1 microsecond is 2000 TSC, so starting at range 5 or 6 is probably OK. We don't need precision on anything lower than that.

So now instead of 58 ranges (0 through 57 inclusive), we have 25 ranges (6 through 30 inclusive). That cuts the memory usage in half. Of course the real savings would be cutting down on that default 7 shift value, which changes my math here slightly, but I think you get the idea.

-Jim

**Comment by SWaoter** [comment]
> I am OK with that. Instead of parameter for num ranges though, I would suggest specifying a min and max range index.

I agree that ranges suites better here. Do you have any plans on this? If not I can work on expanding histogram/bdev_enable_histogram API and after that we can discuss further changes if needed.

**Comment by jimharris** [comment]
> > I am OK with that. Instead of parameter for num ranges though, I would suggest specifying a min and max range index.
> 
> I agree that ranges suites better here. Do you have any plans on this? If not I can work on expanding histogram/bdev_enable_histogram API and after that we can discuss further changes if needed.

I have no plans to work on this directly but would happy to discuss further here, or Slack, or Gerrit code reviews.

**Comment by jimharris** [comment]
https://review.spdk.io/c/spdk/spdk/+/25383

**Comment by SWaoter** [comment]
Hi @jimharris and @shuhei-matsumoto,
I see that patches related to shrinking histogram size are actively reviewed. Can we discuss further possible improvements?
We will have enabling API:
`bdev_enable_histogram --name test --min-nsec 1000 --max-nsec 1000000000` which is quite verbose for user. However getting histogram values will still be possible only in base64 format that later should be parsed.
What do you think about original idea of extending `bdev_get_histogram` to allow providing user defined buckets? Then when such option specified we will return not entire histogram in base64, but only distribution across specified bucket.
So the process will be following:
`bdev_enable_histogram --name test --min-nsec 1000 --max-nsec 1000000000`

`bdev_get_histogram --name test --buckets-nsec '10000 20000 30000 40000 50000 100000 500000 10000000 1000000000'` // that will transform to [0,10us], [10us, 20us], [20us, 30us], [30us, 40us], [40us, 50us], [50us,100us], [100us, 500us], [500us, 10ms], [10ms, 1s], [1s, inf] buckets.
The result of the call will be array that represent histogram distribution across user-buckets.

We will be able to specify more buckets for sweet spot of our device with fewer buckets on the tail.
Such format is useful as it later can be easily stored in [prometheus format](https://prometheus.io/docs/tutorials/understanding_metric_types/#histogram) and reported to dashboards.
Under the hood in `bdev_get_histogram` we will just do `spdk_histogram_data_iterate` and dump internal histogram buckets to user-provided ones.

**Comment by SWaoter** [comment]
> It requires that the unit of datapoints should be not raw TSC count but nanoseconds/microseconds?

No, we still can save all datapoints in TSC. Whole path will look something like this:
1) User call bdev_get_histogram --name test --buckets-nsec '...' with nanoseconds buckets
2) At RPC level we translate nanoseconds to TSC
3) Call spdk_histogram_data_iterate and dump histogram to user-defined buckets by comparing TSC borders
4) Return histogram distribution across user-defined buckets

**Comment by shuhei-matsumoto** [comment]
Thank you for your clarification. I see. I already tried similar one. What I did was to modify histogram.py to have custom ranges. Is this enough? But, I think your idea is possible.

**Comment by SWaoter** [comment]
@shuhei-matsumoto,
In my usecase I want to export SPDK bdev metrics in Prometheus format on HTTP endpoint. It is an overkill to do it inside SPDK app, so I have sidecar that periodically poll SPDK app (by triggering bdev_get_iostat and bdev_histogram) to obtain stats and expose it in needed format.
So while it is possible to use histogram.py to parse histogram, it still would be better to get all needed data directly from JSON-RPC API.

**Comment by shuhei-matsumoto** [comment]
@SWaoter Thanks for your further clarification. I see. bdev_enable_histogram will get nanoseconds parameters. Yours looks nice and reasonable enhancement to me. I am happy to review if you work on. @jimharris may have better or different feedback. Please wait.

**Comment by jimharris** [comment]
All of this could be done by integrating histogram.py capabilities into rpc.py. We don't have to do it in the C app itself.

The C app should provide the raw data and allow applications process as desired. Putting this in the C code just invites feature creep to put more of this kind of analysis work in the C app itself.

rpc.py could take the buckets as a parameter to bdev_enable_histogram RPC, and configure the granularity and range appropriately - to both reduce memory usage and the amount of data that needs to be transferred over RPC.

**Comment by SWaoter** [comment]
> All of this could be done by integrating histogram.py capabilities into rpc.py. We don't have to do it in the C app itself.

Although it can be done in rpc.py it will mean that all applications that communicate with SPDK directly through socket will have to duplicate same functionality.

**Comment by tomzawadzki** [comment]
[Bug scrub] What is the latest on this enhancement, had the merged patch resolve all the concerns ? https://review.spdk.io/c/spdk/spdk/+/25384/14

**Comment by SWaoter** [comment]
Hi, @tomzawadzki . Thanks for bringing it up. There are indeed some improvements in histogram part, but the main point of original issue still remain.

The thing is that even though we've done some work to reduce histogram internal representation size and added some flexibility here, I still think that the way how it can be obtained from SPDK is inadequate.

We have to obtain base64 histogram representation(which is tens of megabytes) and later parse this as a client.
I would still prefer my original approach when we would've added possibility to pass some intervals to bdev_get_histogram, so the app will dump histogram in this provided buckets. It can be done in `_rpc_bdev_histogram_data_cb` and it's pretty simple with `spdk_histogram_data_iterate`.

However since this proposed solution was not approved, I hadn't worked on it. If the opinion on this has changed let me know.
